<a href="https://colab.research.google.com/github/GuJaco/Aula01CienciadeDados/blob/main/Tarefa_05_06.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [19]:
import pandas as pd

# Carregar os dados
dados = pd.read_csv("https://raw.githubusercontent.com/GuJaco/Aula01CienciadeDados/refs/heads/main/operators.csv")

# Exibir as primeiras linhas do conjunto de dados
dados


Organization        Name      Role  Armor  Speed  Difficulty  \
0           APCA       AMARU  ATTACKER      2      2           2   
1           APCA        GOYO  DEFENDER      2      2           2   
2   JAEGER CORPS        NOKK  ATTACKER      2      2           3   
3   JAEGER CORPS      WARDEN  DEFENDER      3      1           2   
4           SASR      MOZZIE  DEFENDER      2      2           2   
5           SASR    GRIDLOCK  ATTACKER      3      1           1   
6           GIGR       NOMAD  ATTACKER      2      2           3   
7           GIGR        KAID  DEFENDER      3      1           2   
8            MPS       CLASH  DEFENDER      3      1           3   
9            MPS    MAVERICK  ATTACKER      1      3           2   
10        G.I.S.     MAESTRO  DEFENDER      3      1           2   
11        G.I.S.       ALIBI  DEFENDER      1      3           3   
12          GIGN        LION  ATTACKER      2      2           1   
13          GIGN       FINKA  ATTACKER      2      2           1   
14     707TH SMB       VIGIL  DEFENDER      1      3           3   
15     707TH SMB    DOKKAEBI  ATTACKER      2      2           2   
16      G.R.O.M.       ZOFIA  ATTACKER      2      2           3   
17      G.R.O.M.         ELA  DEFENDER      1      3           1   
18        S.D.U.        YING  ATTACKER      2      2           2   
19        S.D.U.      LESION  DEFENDER      2      2           1   
20        G.E.O.        MIRA  DEFENDER      3      1           3   
21        G.E.O.      JACKAL  ATTACKER      2      2           3   
22        S.A.T.      HIBANA  ATTACKER      1      3           1   
23        S.A.T.        ECHO  DEFENDER      3      1           3   
24          BOPE     CAVEIRA  DEFENDER      1      3           3   
25          BOPE     CAPITAO  ATTACKER      1      3           2   
26     NAVY SEAL  BLACKBEARD  ATTACKER      2      2           1   
27     NAVY SEAL    VALKYRIE  DEFENDER      2      2           2   
28          JTF2        BUCK  ATTACKER      2      2           1   
29          JTF2       FROST  DEFENDER      2      2           1   
30           SAS        MUTE  DEFENDER      2      2           1   
31           SAS      SLEDGE  ATTACKER      2      2           1   
32           SAS       SMOKE  DEFENDER      2      2           2   
33           SAS    THATCHER  ATTACKER      2      2           1   
34          SWAT         ASH  ATTACKER      1      3           2   
35          SWAT      CASTLE  DEFENDER      2      2           2   
36          SWAT       PULSE  DEFENDER      1      3           3   
37          SWAT    THERMITE  ATTACKER      2      2           1   
38          GIGN    MONTAGNE  ATTACKER      3      1           3   
39          GIGN      TWITCH  ATTACKER      2      2           2   
40          GIGN         DOC  DEFENDER      3      1           1   
41          GIGN        ROOK  DEFENDER      3      1           1   
42          GSG9       JAGER  DEFENDER      1      3           1   
43         GSG10      BANDIT  DEFENDER      1      3           1   
44         GSG11       BLITZ  ATTACKER      2      2           3   
45         GSG12          IQ  ATTACKER      1      3           3   
46      SPETSNAZ        FUZE  ATTACKER      3      1           1   
47      SPETSNAZ        GLAZ  ATTACKER      2      2           2   
48      SPETSNAZ    TACHANKA  DEFENDER      3      1           1   
49      SPETSNAZ      KAPKAN  DEFENDER      2      2           1   
50    NIGHTHAVEN        KALI  ATTACKER      2      2           2   
51    NIGHTHAVEN       WAMAI  DEFENDER      2      2           2   
52           REU        IANA  ATTACKER      1      2           2   
53  UNAFFILIATED        ORYX  DEFENDER      2      2           2   

               Primary1        Primary2 Primary3         Secondary1  \
0                  G8A1       SUPERNOVA      NaN             ITA12S   
1        VECTOR .45 ACP          TCSG12      NaN               P229   
2                 FMG-9        SIX12 SD      NaN  

In [23]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Carregar os dados
url = "https://raw.githubusercontent.com/GuJaco/Aula01CienciadeDados/refs/heads/main/operators.csv"
dados = pd.read_csv(url)

# Definir a variável-alvo
y = dados["Role"]  # Classificando entre ATTACKER e DEFENDER

# Selecionar as variáveis preditoras (removendo colunas irrelevantes ou com muitos valores não numéricos)
X = dados[['Armor', 'Speed', 'Difficulty']]  # Mantendo apenas colunas numéricas

# Converter a variável alvo para valores numéricos
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)  # Converte ATTACKER e DEFENDER para 0 e 1

# Dividir o dataset em conjunto de treinamento e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Criar e treinar o modelo Naive Bayes
modelo = GaussianNB()
modelo.fit(X_train, y_train)

# Fazer previsões
y_pred = modelo.predict(X_test)

# Avaliar o modelo
acuracia = accuracy_score(y_test, y_pred)
matriz_confusao = confusion_matrix(y_test, y_pred)
relatorio_classificacao = classification_report(y_test, y_pred)

# Exibir os resultados
print(f"Acurácia: {acuracia}")
print("Matriz de Confusão:")
print(matriz_confusao)
print("Relatório de Classificação:")
print(relatorio_classificacao)


Acurácia: 0.6470588235294118
Matriz de Confusão:
[[7 2]
 [4 4]]
Relatório de Classificação:
              precision    recall  f1-score   support

           0       0.64      0.78      0.70         9
           1       0.67      0.50      0.57         8

    accuracy                           0.65        17
   macro avg       0.65      0.64      0.64        17
weighted avg       0.65      0.65      0.64        17



Conclusao

O modelo Naive Bayes obteve 64,7% de acurácia, classificando bem os defensores (78% de recall), mas com dificuldades nos atacantes (50% de recall). A inclusão de mais variáveis (armas, gadgets) e o balanceamento dos dados podem melhorar o desempenho

In [24]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Carregar os dados
dados = pd.read_csv("https://raw.githubusercontent.com/GuJaco/Aula01CienciadeDados/refs/heads/main/operators.csv")

# Selecionar features numéricas e target (assumindo que 'Role' seja a variável alvo)
dados['Role'] = dados['Role'].map({'ATTACKER': 1, 'DEFENDER': 0})  # Convertendo para valores numéricos
X = dados[['Armor', 'Speed', 'Difficulty']]
y = dados['Role']

# Divisão em treino e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Padronização dos dados
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Treinar modelo K-NN (k=5 como exemplo)
k = 5
modelo_knn = KNeighborsClassifier(n_neighbors=k)
modelo_knn.fit(X_train, y_train)

# Fazer previsões
y_pred = modelo_knn.predict(X_test)

# Avaliação do modelo
acuracia = accuracy_score(y_test, y_pred)
matriz_confusao = confusion_matrix(y_test, y_pred)
relatorio_classificacao = classification_report(y_test, y_pred)

# Exibir resultados
print(f"Acurácia: {acuracia:.2f}")
print("\nMatriz de Confusão:")
print(matriz_confusao)
print("\nRelatório de Classificação:")
print(relatorio_classificacao)


Acurácia: 0.55

Matriz de Confusão:
[[2 4]
 [1 4]]

Relatório de Classificação:
              precision    recall  f1-score   support

           0       0.67      0.33      0.44         6
           1       0.50      0.80      0.62         5

    accuracy                           0.55        11
   macro avg       0.58      0.57      0.53        11
weighted avg       0.59      0.55      0.52        11



Conclusão

O modelo K-NN teve 55% de acurácia, errando muitos defensores. Melhorias incluem ajustar K, adicionar variáveis e testar outros modelos para maior precisão.

In [25]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Carregar os dados
dados = pd.read_csv("https://raw.githubusercontent.com/GuJaco/Aula01CienciadeDados/refs/heads/main/operators.csv")

# Selecionar features numéricas e target
dados['Role'] = dados['Role'].map({'ATTACKER': 1, 'DEFENDER': 0})  # Convertendo para valores numéricos
X = dados[['Armor', 'Speed', 'Difficulty']]
y = dados['Role']

# Divisão em treino e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Padronização dos dados
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Modelo Árvore de Decisão
decision_tree = DecisionTreeClassifier(random_state=42)
decision_tree.fit(X_train, y_train)
y_pred_tree = decision_tree.predict(X_test)

# Modelo Random Forest
random_forest = RandomForestClassifier(n_estimators=100, random_state=42)
random_forest.fit(X_train, y_train)
y_pred_rf = random_forest.predict(X_test)

# Avaliação dos modelos
def avaliar_modelo(nome, y_test, y_pred):
    print(f"\n=== {nome} ===")
    print(f"Acurácia: {accuracy_score(y_test, y_pred):.2f}")
    print("Matriz de Confusão:")
    print(confusion_matrix(y_test, y_pred))
    print("Relatório de Classificação:")
    print(classification_report(y_test, y_pred))

avaliar_modelo("Árvore de Decisão", y_test, y_pred_tree)
avaliar_modelo("Random Forest", y_test, y_pred_rf)



=== Árvore de Decisão ===
Acurácia: 0.73
Matriz de Confusão:
[[4 2]
 [1 4]]
Relatório de Classificação:
              precision    recall  f1-score   support

           0       0.80      0.67      0.73         6
           1       0.67      0.80      0.73         5

    accuracy                           0.73        11
   macro avg       0.73      0.73      0.73        11
weighted avg       0.74      0.73      0.73        11


=== Random Forest ===
Acurácia: 0.73
Matriz de Confusão:
[[4 2]
 [1 4]]
Relatório de Classificação:
              precision    recall  f1-score   support

           0       0.80      0.67      0.73         6
           1       0.67      0.80      0.73         5

    accuracy                           0.73        11
   macro avg       0.73      0.73      0.73        11
weighted avg       0.74      0.73      0.73        11



Conclusão

Análise dos Modelos de Árvores de Decisão e Random Forest
Ambos os modelos obtiveram 73% de acurácia, mostrando um desempenho equilibrado entre atacantes e defensores.

Árvore de Decisão: Teve 80% de precisão para defensores e 67% para atacantes, com um recall inverso (67% e 80%, respectivamente).

Random Forest: Obteve resultados idênticos, o que pode indicar que o conjunto de dados não está se beneficiando da robustez do modelo.

A matriz de confusão mostra que o modelo comete erros similares em ambas as classes, sugerindo que mais variáveis ou ajustes podem melhorar a classificação.

Conclusão
Os modelos tiveram um desempenho razoável, mas ajustes como mais variáveis, otimização de hiperparâmetros ou aumento dos dados podem trazer ganhos.